{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "20955012",
   "metadata": {},
   "source": [
    "# Aim of the notebook\n",
    "In this notebook we will convert a SpatialData object to an anndata containing sufficient information to perform spatial RNA velocity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0076c18e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7dbc5a1",
   "metadata": {},
   "source": [
    "We define the path where the SpatialData object is stored (as zarr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef3a978d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def xenium_to_adata_for_velocity(path:str='',max_dist_to_nucleus='all'):\n",
    "    \"\"\" Given the path of a Xenium dataset generated by the machine, format it to adata\n",
    "    \n",
    "    \"\"\"\n",
    "    #read adata\n",
    "    adata=sc.read_10x_h5(path+'/cell_feature_matrix.h5')\n",
    "    adata.var = adata.var.rename(columns={\"gene_ids\":\"Ensemble ID\"})\n",
    "    adata.var.reset_index(inplace=True, drop = False, names= \"gene_name\")\n",
    "    adata.obs.reset_index(inplace=True, drop=False, names= \"cell_id\")\n",
    "    df_cells = pd.read_csv(path+\"/cells.csv.gz\")\n",
    "    df_adata_obs = df_cells[['cell_id','x_centroid',\"y_centroid\", \"transcript_counts\", \"total_counts\", \"cell_area\", \"nucleus_area\"]]\n",
    "    adata.obs = df_adata_obs.copy()\n",
    "    del df_adata_obs\n",
    "    del df_cells\n",
    "    #read transcripts\n",
    "    transcripts = pd.read_csv(path+\"/transcripts.csv.gz\")\n",
    "    transcripts[\"cell_mapped\"] = transcripts[\"cell_id\"].apply(lambda x: \"unassigned\" if x == \"UNASSIGNED\" else \"assigned\")\n",
    "    # check proportion of transcripts assigned to a cell \n",
    "    print(transcripts[\"cell_mapped\"].value_counts())\n",
    "    print('Percentage of unassigned_transcripts',np.round((transcripts[\"cell_mapped\"].value_counts()[1]/len(transcripts))*100,1), \"%\")\n",
    "    tr = transcripts[transcripts[\"cell_id\"] != \"UNASSIGNED\" ]\n",
    "    tr.reset_index(inplace=True, drop=True)\n",
    "    # keep only counts situated at less than a certain distance to the nuclei edge\n",
    "    if max_dist_to_nucleus!='all':\n",
    "        tr=tr[tr['nucleus_distance']<max_dist_to_nucleus]\n",
    "    \n",
    "    # divide data in nuclei and cytopasm\n",
    "    tn = tr[tr['overlaps_nucleus'] == 1]\n",
    "    tc = tr[tr['overlaps_nucleus'] == 0]\n",
    "\n",
    "    # create the cellxgene matrix for nuc counts and cytoplasmic\n",
    "    nuc = pd.crosstab(tn['cell_id'],tn['feature_name'])\n",
    "    cyt = pd.crosstab(tc['cell_id'],tc['feature_name'])\n",
    "\n",
    "    # filer adata to keep only common cells \n",
    "    cell_tr = pd.crosstab(tr['cell_id'],tr['feature_name'])\n",
    "    adata = adata[adata.obs['cell_id'].isin(cell_tr.index)]\n",
    "    adata.obs.index = adata.obs.index.astype(str)\n",
    "    \n",
    "    # eliminate genes  that are not present in nuc and cyt\n",
    "    adata.obs.index = adata.obs.index.astype(str)\n",
    "    adata.var.index = adata.var.index.astype(str)\n",
    "    adata = adata[:,adata.var['gene_name'].isin(nuc.columns)]\n",
    "    adata.obs.index = adata.obs.index.astype(str)\n",
    "    adata.var.index = adata.var.index.astype(str)\n",
    "    adata = adata[:,adata.var['gene_name'].isin(cyt.columns)]\n",
    "    adata.obs.index = adata.obs.index.astype(str)\n",
    "    adata.var.index = adata.var.index.astype(str)\n",
    "    nuc.index=nuc.index.astype(str)\n",
    "    cyt.index=cyt.index.astype(str)\n",
    "    \n",
    "    #get the cells which has a transcripts mappep to it \n",
    "    adata = adata[adata.obs['cell_id'].isin(nuc.index)]\n",
    "    adata.obs.index = adata.obs.index.astype(str)\n",
    "    adata.var.index = adata.var.index.astype(str)\n",
    "    adata = adata[adata.obs['cell_id'].isin(cyt.index)]\n",
    "    \n",
    "    # create \"spliced\", \"unspliced\" layers since scvelo looks for them\n",
    "    nucsort = nuc.loc[adata.obs['cell_id'],adata.var[\"gene_name\"]]\n",
    "    cytsort = cyt.loc[adata.obs['cell_id'],adata.var[\"gene_name\"]]\n",
    "\n",
    "    # define nuclear and cytoplasmic counts in layers. Nuclear counts are stores in 'unspliced' and cytoplasmic counts are stored in 'spliced'\n",
    "    adata.layers['spliced'] = np.array(cytsort)\n",
    "    adata.layers['unspliced'] = np.array(nucsort)\n",
    "\n",
    "    # the counts of all cells are considered the sum of spliced and unspliced\n",
    "    adata.X=adata.layers['spliced']+adata.layers['unspliced']\n",
    "    adata.obsm['spatial']=np.array(adata.obs.loc[:,['x_centroid','y_centroid']])\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "142a25a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_default_X(adata:'Anndata',to='all'):\n",
    "    ''' Set the default expression matrix to either 'all', the sum of spliced and spliced, 'spliced' or 'unspliced'\n",
    "     Parameters\n",
    "    ----------\n",
    "    adata:'AnnData object'\n",
    "        Adata object including cell expression, unspiced and spliced matrices\n",
    "    to:'str'\n",
    "        Expression matrix to set up as a default in .X. It can be ither 'all' for the sum of expression and spliced,\n",
    "        'unspliced' or 'spliced' for using only part of the counts\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    adata:'AnnData object'\n",
    "        Adata object with redefined X    \n",
    "    '''\n",
    "    \n",
    "    if to=='all':\n",
    "        adata.X=adata.layers['spliced']+adata.layers['unspliced']\n",
    "    if to=='spliced':\n",
    "        adata.X=adata.layers['spliced']\n",
    "    if to=='unspliced':\n",
    "        adata.X=adata.layers['unspliced']\n",
    "    \n",
    "    return adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5e8f06",
   "metadata": {},
   "source": [
    "# Define paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f1720719",
   "metadata": {},
   "outputs": [],
   "source": [
    "path='/media/sergio/Meninges/meninges/20230705__20230705_MENINGES_ELIN_run2/output-XETG00047__0005264__HE27a__20230705__134603'\n",
    "path_to_write='/media/sergio/Meninges/SpatialData_formatted/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14fd9d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata=xenium_to_adata_for_velocity(path:str='',max_dist_to_nucleus=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "41de8d24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 286581 × 389\n",
       "    obs: 'cell_id', 'x_centroid', 'y_centroid', 'transcript_counts', 'total_counts', 'cell_area', 'nucleus_area'\n",
       "    var: 'gene_name', 'Ensemble ID', 'feature_types', 'genome'\n",
       "    layers: 'spliced', 'unspliced'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata=set_default_X(adata,to='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "738e53fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write(path_to_write+'adata_for_velocity.h5ad')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SpatialData2",
   "language": "python",
   "name": "spatialdata2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
